core/src/main/scala/org/apache/spark/rdd/OrderedRDDFunctions.scala

```diff
@@ -73,7 +75,23 @@ class OrderedRDDFunctions[K : Ordering : ClassTag,
    * because it can push the sorting down into the shuffle machinery.
    */
   def repartitionAndSortWithinPartitions(partitioner: Partitioner): RDD[(K, V)] = self.withScope {
-    new ShuffledRDD[K, V, V](self, partitioner).setKeyOrdering(ordering)
+    if (self.partitioner == Some(partitioner)) {
+      self.mapPartitions(iter => {
+        val context = TaskContext.get
+        val sorter = new ExternalSorter[K, V, V](context, None, None, Some(ordering))
+        sorter.insertAll(iter)
+        context.taskMetrics.incMemoryBytesSpilled(sorter.memoryBytesSpilled)
+        context.taskMetrics.incDiskBytesSpilled(sorter.diskBytesSpilled)
```

Comment:
These two may not accurate since the `sorter.iterator`  could also spill during traverse?

Although I also see other places use it in the same way.

Reply:
I review the related codes and it seems that `sorter.iterator` may spill during traverse:
`isShuffleSort = false` in `def iterator` makes internals iterator `destructiveIterator` a `SpillableIterator`.



We can add the update of `taskMetrics` to a task completion listener if necessary, but maybe in a new ticket.

As to this PR, I perfer to keep the line with existing impl.

Reply:
SGTM